diff --git a/src/cffi-abcl.lisp b/src/cffi-abcl.lisp
index f01bbd7..022e6aa 100644
--- a/src/cffi-abcl.lisp
+++ b/src/cffi-abcl.lisp
@@ -527,7 +527,7 @@ (defmacro %defcallback (name return-type arg-names arg-types body
          (jinterface-implementation
           (ensure-callback-interface ',return-type ',arg-types)
           "callback"
-          (convert-args-to-lisp-values ,arg-names ,arg-types ,body))))
+          (convert-args-to-lisp-values ,arg-names ,arg-types (lisp-value-to-java ,body ',return-type)))))
 ;;          (lambda (,@arg-names) ,body))))
 
 (jvm::define-class-name +callback-object+ "com.sun.jna.Callback")
diff --git a/tests/callbacks.lisp b/tests/callbacks.lisp
index 9faf4dd..94da57c 100644
--- a/tests/callbacks.lisp
+++ b/tests/callbacks.lisp
@@ -75,6 +75,7 @@ (defcallback sum-unsigned-int :unsigned-int
   (+ a b))
 
 (defcallback sum-long :long ((a :long) (b :long))
+  (format T "42 ~X ~X~%" a b)
   (+ a b))
 
 (defcallback sum-unsigned-long :unsigned-long
diff --git a/tests/libtest.c b/tests/libtest.c
index 0f0e61b..fa2dcae 100644
--- a/tests/libtest.c
+++ b/tests/libtest.c
@@ -165,6 +165,7 @@ int expect_unsigned_int_sum(unsigned int (*f)(unsigned int, unsigned int))
 DLLEXPORT
 int expect_long_sum(long (*f)(long, long))
 {
+    printf("\n>>> LONG: %lx <<<\n", f(LONG_MIN+1, -1));
     return f(LONG_MIN+1, -1) == LONG_MIN;
 }
 
